x-pack/plugins/fleet/server/services/epm/packages/install_steps.ts

```diff
@@ -0,0 +1,451 @@
+
+  const { packageInfo } = packageInstallContext;
+
+  const packageAssetResults = await withPackageSpan('Update archive entries', () =>
+    saveArchiveEntriesFromAssetsMap({
+      savedObjectsClient,
+      assetsMap: packageInstallContext.assetsMap,
+      paths: packageInstallContext.paths,
+      packageInfo,
+      installSource,
+    })
+  );
+  const packageAssetRefs: PackageAssetReference[] = packageAssetResults.saved_objects.map(
+    (result) => ({
+      id: result.id,
+      type: ASSETS_SAVED_OBJECT_TYPE,
+    })
+  );
+
+  return { packageAssetRefs, installedKibanaAssetsRefs };
+}
+
+export async function stepSaveSystemObject(context: InstallContext) {
+  const {
+    packageInstallContext,
+    savedObjectsClient,
+    logger,
+    esClient,
+    installedPkg,
+    packageAssetRefs,
+  } = context;
+  const { packageInfo } = packageInstallContext;
+  const { name: pkgName, version: pkgVersion } = packageInfo;
+
+  auditLoggingService.writeCustomSoAuditLog({
+    action: 'update',
+    id: pkgName,
+    savedObjectType: PACKAGES_SAVED_OBJECT_TYPE,
+  });
+
+  await withPackageSpan('Update install status', () =>
+    savedObjectsClient.update<Installation>(PACKAGES_SAVED_OBJECT_TYPE, pkgName, {
+      version: pkgVersion,
+      install_version: pkgVersion,
+      install_status: 'installed',
+      package_assets: packageAssetRefs,
+      install_format_schema_version: FLEET_INSTALL_FORMAT_VERSION,
+      latest_install_failed_attempts: clearLatestFailedAttempts(
+        pkgVersion,
+        installedPkg?.attributes.latest_install_failed_attempts ?? []
+      ),
+    })
+  );
+
+  // Need to refetch the installation again to retrieve all the attributes
+  const updatedPackage = await savedObjectsClient.get<Installation>(
+    PACKAGES_SAVED_OBJECT_TYPE,
+    pkgName
+  );
+  logger.debug(`Package install - Install status ${updatedPackage?.attributes?.install_status}`);
+  // If the package is flagged with the `keep_policies_up_to_date` flag, upgrade its
+  // associated package policies after installation
+  if (updatedPackage.attributes.keep_policies_up_to_date) {
+    await withPackageSpan('Upgrade package policies', async () => {
+      const policyIdsToUpgrade = await packagePolicyService.listIds(savedObjectsClient, {
+        page: 1,
+        perPage: SO_SEARCH_LIMIT,
+        kuery: `${PACKAGE_POLICY_SAVED_OBJECT_TYPE}.package.name:${pkgName}`,
+      });
+      logger.debug(
+        `Package install - Package is flagged with keep_policies_up_to_date, upgrading its associated package policies ${policyIdsToUpgrade}`
+      );
+      await packagePolicyService.upgrade(savedObjectsClient, esClient, policyIdsToUpgrade.items);
+    });
+  }
+  logger.debug(
+    `Install status ${updatedPackage?.attributes?.install_status} - Installation complete!`
+  );
+}
+
+// Function invoked after each transition
+export const updateLatestExecutedState = async (context: InstallContext) => {
```

Comment:
> Well all this function does is to write the latest state to the SO. If a ConcurrentInstallError happens, I think that it would be good to save this error as reason of the last failure. Is there a specific reason why we wouldn't like to save it?

If there is a `ConcurrentInstallError` this mean another Kibana instance is trying to run the package install so we should avoid writing the state from kibana instances that are not running the install, does this make sense to you?

Reply:
I see, so in this case we can add an exception for this specific error as you were suggesting above. That means that we'll never report when an error was of this type though.

Reply:
> I see, so in this case we can add an exception for this specific error as you were suggesting above. That means that we'll never report when an error was of this type though.

Yes it seems fine to me to not those errors as they are designed for that

Reply:
@criamico did you catch that `ConcurrentInstallOperationError` somewhere I was not able to find it? Also I think it is a case that may need a test to be sure we do not break that feature, as we will almost never get it locally running one Kibana instance, but that could happens frequently in cloud or in any scenario with multiple Kibana

Reply:
@nchaulet no, you're right is not there, I'm still working on that part. I fixed some other bugs that I found through tests so that's should be the only outstanding part that I'm missing. I'll also add tests to cover it as you suggested.